### ISSUE TITLE
No Content-Type found for MIME entity

### ISSUE THREAD

**Issue by kalbasit** [jordan-wright/email#28]
Parsing the following email (pasted as is)

```
From: Mikhail Gusarov <dottedmag@dottedmag.net>
To: notmuch@notmuchmail.org
References: <20091117190054.GU3165@dottiness.seas.harvard.edu>
Date: Wed, 18 Nov 2009 01:02:38 +0600
In-Reply-To: <20091117190054.GU3165@dottiness.seas.harvard.edu> (Lars
	Kellogg-Stedman's message of "Tue, 17 Nov 2009 14:00:54 -0500")
Message-ID: <87iqd9rn3l.fsf@vertex.dottedmag>
User-Agent: Gnus/5.13 (Gnus v5.13) Emacs/23.1 (gnu/linux)
MIME-Version: 1.0
Subject: Re: [notmuch] Working with Maildir storage?
X-BeenThere: notmuch@notmuchmail.org
X-Mailman-Version: 2.1.12
Precedence: list
List-Id: "Use and development of the notmuch mail system."
	<notmuch.notmuchmail.org>
List-Unsubscribe: <http://notmuchmail.org/mailman/options/notmuch>,
	<mailto:notmuch-request@notmuchmail.org?subject=unsubscribe>
List-Archive: <http://notmuchmail.org/pipermail/notmuch>
List-Post: <mailto:notmuch@notmuchmail.org>
List-Help: <mailto:notmuch-request@notmuchmail.org?subject=help>
List-Subscribe: <http://notmuchmail.org/mailman/listinfo/notmuch>,
	<mailto:notmuch-request@notmuchmail.org?subject=subscribe>
Content-Type: multipart/mixed; boundary="===============1958295626=="
Sender: notmuch-bounces@notmuchmail.org
Errors-To: notmuch-bounces@notmuchmail.org

--===============1958295626==
Content-Type: multipart/signed; boundary="=-=-=";
	micalg=pgp-sha1; protocol="application/pgp-signature"

--=-=-=
Content-Transfer-Encoding: quoted-printable


Twas brillig at 14:00:54 17.11.2009 UTC-05 when lars@seas.harvard.edu did g=
yre and gimble:

 LK> Resulted in 4604 lines of errors along the lines of:

 LK>   Error opening
 LK>   /home/lars/Mail/read-messages.2008/cur/1246413773.24928_27334.hostna=
me,U=3D3026:2,S:
 LK>   Too many open files

See the patch just posted here.

=2D-=20
  http://fossarchy.blogspot.com/

--=-=-=
Content-Type: application/pgp-signature

-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1.4.9 (GNU/Linux)

iQIcBAEBAgAGBQJLAvNOAAoJEJ0g9lA+M4iIjLYQAKp0PXEgl3JMOEBisH52AsIK
CzzfP4Fzd41K9VH/c1EdQWDYR6FCAA4IUSNICnJhITsYUb0eC5AKJiey3JP0+rmd
s4qEFBKH2iuphv8Llltcv2Q8DyPuJBkVa3mO9XCCeABZ6v4UvnTSWRVG12csSEih
ScgienU8sMrM9LwvvVI1ZB2flm2TzsH2hWi30jIgmtBntIKJaTgbFXB50FYFwULa
gGL/oH3u+YpumedWzPZdCJrw2q7nMvYx8aQ29EDCNLZibAZe+6oDTa6Fv6/0ldpQ
U+DptR0nJGbJTWa26OTSvmyeIORjAfM+TEI68n7KO9VHYPmVh6awcf0MNKYh2xWk
eRQNBcKyQNWxeKyCCpT/rrTlpxBWahpvg+V8lkDH2W09wjRp6CUKvifK3Sz3am9m
5ZUMpvXbwkZD6Ci6l/QytbYK50e8UpvFSu5DBaxBz59ykoypuNg2ayO5Kdi6IF5d
T+Sw6wo8UKn9a33+vheIc0fkhZXbeSotEmDm7huazm6CgM3dcWXUpTuJvik1cSWp
4buv98gY6IKWKoUTXODWUr+7VR4gei8du8qOsKem+QDfNX7tmaIRjhrbB24B91Wy
td3MTJD7GjMNid0INqRY1CRMLo8YlPaq6NBZfcYtYgwa6gpJijz1/MAn8+GMrfhF
9LI8b9jopNP+pMYBohLA
=/ksP
-----END PGP SIGNATURE-----
--=-=-=--

--===============1958295626==
Content-Type: text/plain; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: inline

_______________________________________________
notmuch mailing list
notmuch@notmuchmail.org
http://notmuchmail.org/mailman/listinfo/notmuch

--===============1958295626==--

```

From gmuch itself:

```
kalbasit@cratos ~/code/src/github.com/gmuch/gmuch [email_bug] ± % make test
=== RUN   TestQuery
--- PASS: TestQuery (0.01s)
=== RUN   TestThreadIDValidation
method=getMessage messageID=87iqd9rn3l.fsf@vertex.dottedmag operation="parse email file" path=/Users/kalbasit/code/src/github.com/gmuch/gmuch/fixtures/database-v1/new/04:2, error="No Content-Type found for MIME entity"
--- FAIL: TestThreadIDValidation (0.00s)
        thread_test.go:25: Thread("0000000000000001"): want error %!q(<nil>) got "No Content-Type found for MIME entity"
=== RUN   TestThread
method=getMessage messageID=87iqd9rn3l.fsf@vertex.dottedmag operation="parse email file" path=/Users/kalbasit/code/src/github.com/gmuch/gmuch/fixtures/database-v1/new/04:2, error="No Content-Type found for MIME entity"
--- FAIL: TestThread (0.03s)
        thread_test.go:40: g.Thread("0000000000000001"): got error: No Content-Type found for MIME entity
FAIL
coverage: 87.7% of statements
exit status 1
FAIL    github.com/gmuch/gmuch  0.064s
make: *** [test] Error 1
```

**Comment by jordan-wright** [comment]
Interesting. And you're sure to be using the latest version of the library? I'll see if I can track this down later.

Thanks for the report!

**Comment by jordan-wright** [comment]
You get the most interesting emails :smile: 

That's a good thing, because I want this to be a comprehensive library that *just works*. In this case, the root cause is that the first subpart (```Content-Transfer-Encoding: quoted-printable```) has not Content-Type specified.

Now, I figured this would be caught by the checks I put in place before, but it doesn't look like that's happening. I'll see if I can get that fixed.

My question to you is - what would you like the ```email.Email``` struct to look like with this message? I see two what I would consider "text" parts, and one PGP signature - not sure where to put that. What would be the best way to handle these two text parts? One of them has to go in ```Email.Text```, but I'm not sure which it should be. Any suggestions?

**Comment by kalbasit** [comment]
> You get the most interesting emails :smile:
lol! Well, at some point I want to port notmuchmail to Go so I am using their test database to drive mine. Gmuch is an HTTP/RPC API on top of Notmuch but at some point I will write my own Notmuch clone. The email database is found here: https://github.com/notmuch/notmuch/blob/master/test/test-databases/Makefile.local#L3 (follow the Makefile).

For quoted printable, perhaps look at https://godoc.org/mime/quotedprintable

> My question to you is - what would you like the email.Email struct to look like with this message? I see two what I would consider "text" parts, and one PGP signature - not sure where to put that. What would be the best way to handle these two text parts? One of them has to go in Email.Text, but I'm not sure which it should be. Any suggestions?

I don't feel that this library should handle GPG but should at least provided. So perhaps add two new fields `Email.GPGSignature` and `Email.GPGEncrypted` so we can deal with them somewhere else.

**Comment by jordan-wright** [comment]
> The email database is found here: https://github.com/notmuch/notmuch/blob/master/test/test-databases/Makefile.local#L3 (follow the Makefile)

Thanks! I'll see what I can do with this.

> For quoted printable, perhaps look at https://godoc.org/mime/quotedprintable

The quoted printable part isn't my worry here. By reading a MIME part using a multipart.NewReader, we get the benefit of the [automatic quoted-printable decoding](https://golang.org/src/mime/multipart/multipart.go?#L113) introduced in v1.5. My question was really about which text part out of the two you felt should be the contents of ```Email.Text```. The text are really two separate parts with unique content (even though one looks like a signature), so it'd be hard to determine if one is the primary text content over the other.

> I don't feel that this library should handle GPG but should at least provided. So perhaps add two new fields Email.GPGSignature and Email.GPGEncrypted so we can deal with them somewhere else.

That's a good idea. I would likely incorporate in a new version and try to abstract it out a bit. For example, I might just have an ```Email.Signature``` and ```Email.Encrypted```, or something similar. Maybe I could add PGP support where the library can handle the generation of the signature for the user (first, I'll admittedly need to learn more about how that works!).

**Comment by kalbasit** [comment]
> The quoted printable part isn't my worry here. By reading a MIME part using a multipart.NewReader, we get the benefit of the automatic quoted-printable decoding introduced in v1.5.

Oh cool! I missed that.

>  The text are really two separate parts with unique content (even though one looks like a signature), so it'd be hard to determine if one is the primary text content over the other.

Well the signature is `Content-Type: application/pgp-signature` and the signed is `Content-Type: application/pgp-encrypted` and you also have `application/pgp-keys`. See https://www.ietf.org/rfc/rfc2015.txt for more details.


> That's a good idea. I would likely incorporate in a new version and try to abstract it out a bit. For example, I might just have an Email.Signature and Email.Encrypted, or something similar. Maybe I could add PGP support where the library can handle the generation of the signature for the user.

That'd be cool! But again I feel that logic should probably live in a different library (or even in this one but different package) so it can be used independently. It would be nice to have methods on the `Email` object itself though.

> (first, I'll admittedly need to learn more about how that works!).

Maybe this will help https://godoc.org/golang.org/x/crypto/openpgp

**Comment by kalbasit** [comment]
once this is fixed, I'll run your library across all of my emails (currently `261380` emails), this would be a nice benchmark and a very good test case!

**Comment by kalbasit** [comment]
ok here goes. With [this main](https://gist.github.com/kalbasit/85b7853cd730bfaa7c93) I managed to generate [this log](https://gist.github.com/kalbasit/765fcecb6085783a5c94) for you. I guess we should probably open an issue per error. Let me know which email you want first.

**Comment by jordan-wright** [comment]
*Wow!* 

That's fantastic! Not the results... since clearly I have work to do (:smile:)... but the sheer coverage is great! Thanks so much for running this!

I'll need to see if I can download my corpus of email from gmail and do the same thing.

As for the "No Content-Type found", let me knock out the one we have here and we can re-run for corpus. As far as the others, yeah, we can absolutely look to make some issues with some test cases and see if it's my library or the mail client who's in the wrong.

Also, I noticed your ```main.go``` was done serially. If you want to parse those even faster, you can use a ```sync.WaitGroup``` to run those in goroutines so you have a bunch of emails being parsed in parallel. That'd be **way** faster.

Thanks again for all your help!

**Comment by kalbasit** [comment]
It was intentionally done serially and it I believe the bottleneck is really I/O! Wonderful result...

**Comment by jordan-wright** [comment]
Ok, I think this one particularly has been fixed as of 669ab2f297e91b64bc7940bc8d997e01b87b03cc.

I was checking for the default content type on the root part, but not on subparts before trying to get the content type from them, which failed.

This should (hopefully!) fix a bunch of the "No Content-Type..." logs from your output.

Let me know if this fixes the issue, and we can move forward with a new log to work on! We'll knock em out one at a time :smile:

**Comment by jordan-wright** [comment]
Also, due to the way I am setting ```e.Text```, the **last** text part is what's set (in this case, it's the signature line). I can change this behavior if you think that, in general, the most recent (most relevant) text part will be the first one found. I don't know if the RFC has a spec for it, but I'll check.

**Comment by jordan-wright** [comment]
Whoops! Had a syntax error in my test (teaches me not to commit without running!)

Here's the commit that should fix it for you: f61123ea07e1921ac4f174e8c42225fd39bd6c6a

**Comment by kalbasit** [comment]
> Also, due to the way I am setting e.Text, the last text part is what's set

I think this is undesirable behavior, how about setting the `application/pgp-signature` content-type in the `Signature` field?

**Comment by jordan-wright** [comment]
I could add a ```.Signature``` field, but I'd consider that another issue since it's another feature. Plus, I think we might be talking about two different things :smile:

The "signature" I'm referring to is:

```
_______________________________________________
notmuch mailing list
notmuch@notmuchmail.org
http://notmuchmail.org/mailman/listinfo/notmuch
```

This is what I imagine is the signature in the email (not the PGP signature). This is what's being set in the text field, since I [check for things](https://github.com/jordan-wright/email/blob/master/email.go#L108) with a Content-Type of "text/plain" to set as ```Email.Text```. This skips over the PGP stuff.

Oh, and good news! I was inspired so I'm downloading my entire Gmail corpus as we speak, so I'll have yet another sample set to work with in tracking down issues. Thanks for the great idea! I'll post the results as soon as I have them.

**Comment by kalbasit** [comment]
> The "signature" I'm referring to is:
```_______________________________________________
notmuch mailing list
notmuch@notmuchmail.org
http://notmuchmail.org/mailman/listinfo/notmuch
```

Ah I see. Good question... We could:
- Append the text to the current Text field
- Make the Text a []string

Also I've updated [the main](https://gist.github.com/kalbasit/85b7853cd730bfaa7c93) and [the log](https://gist.github.com/kalbasit/765fcecb6085783a5c94) to reflect the latest version.

Closing this as fixed.

**Comment by jordan-wright** [comment]
Excellent! I'm glad to see that we fixed about 2k issues. I'm definitely ok with that!

So, from what I can tell, there are about 3 unique issues left:

* ```quotedprintable: invalid unescaped byte```
* ```unexpected EOF```
* ```mime: invalid media parameter```
* ```malformed MIME header line:```

Would there be any way you could add test cases for these 4 items? I'd be happy to create the issues for it this evening.

**Comment by kalbasit** [comment]
Sure, I'll push a commit tonight with test cases.

**Comment by jordan-wright** [comment]
Good news! I managed to run a test with about 55k emails I had laying around. Your main worked perfectly, I just adjusted paths, etc. Thanks for that!

[Here's](https://gist.github.com/jordan-wright/6307ed4971d8f178f043) the log that was generated. I'll see if I can generate some test cases out of this.

**Comment by kalbasit** [comment]
Cool! I'm happy that helped. Hopefully we'll get it parsing ALL emails even those not respecting the RFC! I have emails dating back to 2008 from god knows what kind of email client so I feel it's important to parse emails even if they are do not follow the spec verbatim as in reality most people have one or many of those laying around.
